File: Swift/Sources/Configuration/CalculationMethod.swift
//
//  CalculationMethod.swift
//  Adhan
//
//  Copyright © 2018 Batoul Apps. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

import Foundation

/* Preset calculation parameters */
public enum CalculationMethod {

    // Muslim World League
    case muslimWorldLeague

    //Egyptian General Authority of Survey
    case egyptian

    // University of Islamic Sciences, Karachi
    case karachi

    // Umm al-Qura University, Makkah
    case ummAlQura

    // The Gulf Region
    case gulf

    // Moonsighting Committee
    case moonsightingCommittee

    // ISNA
    case northAmerica

    // Kuwait
    case kuwait

    // Qatar
    case qatar

    // Singapore
    case singapore

    // Other
    case other

    public var params: CalculationParameters {
        switch(self) {
        case .muslimWorldLeague:
            return CalculationParameters(fajrAngle: 18, ishaAngle: 17, method: self)
        case .egyptian:
            return CalculationParameters(fajrAngle: 19.5, ishaAngle: 17.5, method: self)
        case .karachi:
            return CalculationParameters(fajrAngle: 18, ishaAngle: 18, method: self)
        case .ummAlQura:
            return CalculationParameters(fajrAngle: 18.5, ishaInterval: 90, method: self)
        case .gulf:
            return CalculationParameters(fajrAngle: 19.5, ishaInterval: 90, method: self)
        case .moonsightingCommittee:
            return CalculationParameters(fajrAngle: 18, ishaAngle: 18, method: self)
        case .northAmerica:
            return CalculationParameters(fajrAngle: 15, ishaAngle: 15, method: self)
        case .kuwait:
            return CalculationParameters(fajrAngle: 18, ishaAngle: 17.5, method: self)
        case .qatar:
            return CalculationParameters(fajrAngle: 18, ishaInterval: 90, method: self)
        case .singapore:
            return CalculationParameters(fajrAngle: 20, ishaAngle: 18, method: self)
        case .other:
            return CalculationParameters(fajrAngle: 0, ishaAngle: 0, method: self)
        }
    }
}

File: src/CalculationMethod.js
import CalculationParameters from './CalculationParameters';

const CalculationMethod = {
    // Muslim World League
    MuslimWorldLeague() {
        const params = new CalculationParameters("MuslimWorldLeague", 18, 17);
        params.methodAdjustments = { dhuhr: 1 };
        return params;
    },

    // Egyptian General Authority of Survey
    Egyptian() {
        const params = new CalculationParameters("Egyptian", 19.5, 17.5);
        params.methodAdjustments = { dhuhr: 1 };
        return params;
    },

    // University of Islamic Sciences, Karachi
    Karachi() {
        const params = new CalculationParameters("Karachi", 18, 18);
        params.methodAdjustments = { dhuhr: 1 };
        return params;
    },

    // Umm al-Qura University, Makkah
    UmmAlQura() {
        return new CalculationParameters("UmmAlQura", 18.5, 0, 90);
    },

    // Dubai
    Dubai() {
        const params = new CalculationParameters("Dubai", 18.2, 18.2);
        params.methodAdjustments = { sunrise: -3, dhuhr: 3, asr: 3, maghrib: 3 };
        return params;
    },

    // Moonsighting Committee
    MoonsightingCommittee() {
        const params = new CalculationParameters("MoonsightingCommittee", 18, 18);
        params.methodAdjustments = { dhuhr: 5, maghrib: 3 };
        return params;
    },

    // ISNA
    NorthAmerica() {
        const params = new CalculationParameters("NorthAmerica", 15, 15);
        params.methodAdjustments = { dhuhr: 1 };
        return params;
    },

    // Kuwait
    Kuwait() {
        return new CalculationParameters("Kuwait", 18, 17.5);
    },

    // Qatar
    Qatar() {
        return new CalculationParameters("Qatar", 18, 0, 90);
    },

    // Singapore
    Singapore() {
        const params = new CalculationParameters("Singapore", 20, 18);
        params.methodAdjustments = { dhuhr: 1 };
        return params;
    },

    // Institute of Geophysics, University of Tehran
    Tehran() {
        const params = new CalculationParameters("Tehran", 17.7, 14, 0, 4.5);
        return params;
    },

    // Dianet
    Turkey() {
        const params = new CalculationParameters("Turkey", 18, 17);
        params.methodAdjustments = { sunrise: -7, dhuhr: 5, asr: 4, maghrib: 7 };
        return params;
    },

    // Other
    Other() {
        return new CalculationParameters("Other", 0, 0);
    }
};

export default CalculationMethod;